### PR TITLE
go: add OAuth2 client credentials authentication to the HTTP transport

### PR DESCRIPTION
**Pull request by rahul-madaan**
The Go counterpart of #4935 (merged), raised separately so neither had to wait on the other. Independent of #4939.

### One-line summary for changelog:
Go client: add OAuth 2.0 client credentials authentication to the HTTP transport.

### Meaningful description

Following the question on #4935 about keeping the supported mechanisms in sync across clients, this adds the OAuth 2.0 client credentials grant ([RFC 6749, section 4.4](https://datatracker.ietf.org/doc/html/rfc6749#section-4.4)) to the Go client. Backends behind an OAuth 2.0 authorization server issue short-lived access tokens to a client ID and secret; the Go client could previously only send a static bearer token, so those backends were unreachable.

Worth noting for that discussion: the clients are already out of sync here. The Go client's `jwt` auth sends a **static** JWT, so it is a duplicate of `apiKey`, and not the token-exchange provider that Python and Java call `jwt`. This PR adds the first auth type in the Go client that fetches a token.

### What this adds

`HTTPAuthConfig` gains `ClientID`, `ClientSecret` and `TokenEndpoint`, plus optional `Scopes`, `ClientAuthMethod` (`client_secret_basic` by default, or `client_secret_post`) and `TokenRefreshBuffer`. The token source is built in `NewWithContext`, so an incomplete or invalid configuration fails when the transport is created rather than on the first event. The type string is `oauth2`, matching the Python and Java clients after the rename in #4935. Exported constants are added for the auth types and client auth methods; the existing `apiKey` and `jwt` values are unchanged.

Token fetching uses `golang.org/x/oauth2/clientcredentials`. **`golang.org/x/oauth2` was already in the module graph** as an indirect dependency of the Google Cloud libraries, so `go.mod` only moves it from the indirect to the direct block at the same version and `go.sum` is unchanged. No new module enters the build.

### Review feedback (thank you, @mobuchowski)

The original token source closed over the context the transport was built with. I reproduced all three consequences against that commit before fixing them:

| Problem | Reproduced |
|---|---|
| Emit ignored its own context for the token request | Emit **hung for 15s** after cancellation |
| Cancelling a `NewWithContext` context broke every later refresh | Emit **failed** with `context canceled` despite a live context |
| `TimeoutInMillis` never applied to token requests | with no client in the context the library falls back to `http.DefaultClient`, which has no timeout |

Tokens are now fetched with the context and HTTP client that `Emit` uses, so the transport's timeout, retries and cancellation all apply, and nothing closes over the construction context. `ReuseTokenSourceWithExpiry` is replaced by a small cache in the token source, which is what makes the per-call context possible, since `oauth2.TokenSource.Token()` takes no context.

### Two further defects found while testing edge cases

Both were reproduced as failing tests first, and both are regression-tested:

- **A concurrent caller ignored its own context.** The cache was guarded by a `sync.Mutex`, and mutex acquisition is not context-aware, so an event waiting on a token request started by another event blocked past its own deadline. This is the same complaint as above, surviving for concurrent callers. The guard is now a one-slot channel with a `select` on `ctx.Done()`.
- **A refresh buffer at or above the token lifetime cost a token request per event.** With the default 120s buffer and a 60s token, every cached token looks stale: three events made three token requests, with no configuration error to warn the user. The buffer is now capped at half the token lifetime, matching the Python client.

### Edge cases verified as already correct, with no code change

A response with no `expires_in` is cached once, following the library's own semantics; an expired token is refetched; malformed, non-JSON and missing-`access_token` responses all surface an error **and send zero unauthenticated lineage requests**; an unusual `token_type` produces the header the library intends; and gzip, custom headers and OAuth coexist.

### Testing

Nine OAuth tests plus subtests: both client auth methods, scope forwarding, token reuse, expiry-triggered refresh, a failing token endpoint, construction-time validation, and one regression test per defect above. The full suite is 74 passing, up from 59 on `main`, green across 10 consecutive fresh runs and 4 under `-race`. `go build`, `go vet` and `golangci-lint` with the repo config are clean, and `go.sum` is unchanged.

Verified against a live Keycloak-protected consumer: both client auth methods authenticated and the consumer accepted the events; a wrong secret produced `oauth2: "unauthorized_client"` with no lineage request attempted; a cancelled context failed fast; and twelve goroutines racing for the very first token produced zero unexpected failures.

The README's HTTP transport example showed an `APIKey` field on `HTTPConfig` that does not exist, so it could not compile; corrected while documenting the new auth type.

### Two things worth your call

- **One unrelated test-only change.** The new tests are parallel, like the rest of the package, and they made a pre-existing flake in `console_test.go` fire far more often: `captureStdout` swaps the process-wide `os.Stdout`, so parallel tests capturing it race. It reproduces on `main` at about 1 run in 8. I serialised `captureStdout` with a mutex so the package is deterministic. Glad to split it out if you would rather.
- **Two pre-existing issues in the Java client that I deliberately did not touch here**, since neither is a regression and fixing them changes merged behaviour: `TokenEndpointTokenProvider` uses `HttpClients.createDefault()`, whose `RequestConfig` sets no response timeout, so a stalled token endpoint hangs the emit; and its refresh buffer is uncapped, giving it the same per-event token request pathology fixed above. Happy to raise an issue for both.

### Checklist
- [x] AI was used in creating this PR

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01QBfMWG5AG33JMqGGQgAV6v

## Summary by CodeRabbit

- **New Features**
  - Added OAuth 2.0 client-credentials authentication for the Go HTTP transport.
  - Supports configurable scopes, token endpoints, client authentication methods, token caching, and automatic refresh before expiration.
  - Preserved API-key and JWT authentication options.
  - Expanded authentication documentation with configuration requirements and examples.

- **Bug Fixes**
  - Serialized console output capture to prevent concurrent capture races.

### CODE REVIEW

**Comment by rahul-madaan** [comment]
@mobuchowski requesting your review.

File: client/go/pkg/transport/http.go [hunk 1]
```diff
@@ -75,10 +118,53 @@ type httpTransport struct {
...
+		Scopes:       auth.Scopes,
+		AuthStyle:    authStyle,
+	}
+
+	refreshBuffer := auth.TokenRefreshBuffer
+	if refreshBuffer <= 0 {
+		refreshBuffer = defaultTokenRefreshBuffer
+	}
+
+	source := tokenSourceFunc(func() (*oauth2.Token, error) { return config.Token(ctx) })
```
Comment (mobuchowski) [inline client/go/pkg/transport/http.go]: This closes over the construction context, so New() makes token requests with context.Background() and OAuth's default HTTP client rather than the Emit context or configured TimeoutInMillis. A stalled token endpoint can therefore block Emit indefinitely despite cancellation; conversely, canceling a context passed to NewWithContext causes every later token refresh to fail even when Emit receives a live context.

**Review by mobuchowski — CHANGES_REQUESTED** [review]

File: client/go/pkg/transport/http.go [hunk 2]
```diff
@@ -75,10 +118,95 @@ type httpTransport struct {
...
+
+	return token, nil
+}
+
+// newClientCredentialsTokenSource builds a token source for the OAuth 2.0 client
+// credentials grant.
+func newClientCredentialsTokenSource(auth *HTTPAuthConfig) (*oauth2TokenSource, error) {
+	if auth.ClientID == "" || auth.ClientSecret == "" || auth.TokenEndpoint == "" {
+		return nil, errors.New("auth type " + AuthTypeOAuth2 +
+			" requires ClientID, ClientSecret and TokenEndpoint")
```
[the remark this answers is not in the document]
Comment (rahul-madaan) [reply client/go/pkg/transport/http.go:178]: The exposure predates this OAuth PR. api_key and jwt put a static bearer token in the Authorization header sent to the same lineage URL, in all three clients, and have done for years.  If plain HTTP is unsafe for an OAuth token, it is equally unsafe for an API key.
